protoc-gen-grpc-gateway/gengateway/template.go

```diff
@@ -208,7 +208,7 @@ var (
 	}
 {{end}}
 
-	return client.{{.Method.GetName}}(ctx, &protoReq)
+	return client.{{.Method.GetName}}(ctx, &protoReq, grpc.Header(&callInfo.HeaderMD), grpc.Trailer(&callInfo.TrailerMD))
```

Comment:
These call options are effective only if `.Method.GetServerStreaming` and `.Method.GetClientStreaming` are false.
The caller of this `request_xxx` function must use `grpc.ClientStream.Header()` and `grpc.ClientStream.Trailer` if otherwise.

IIUC, it would get easier to handle this case if you have let this `request_xxx` function allocate the `metadata.MD`s and let the function return the `MD`s in addition to response and errors.

How about renaming `runtime.CallInfo` to `runtime.ServerMetadata` and declaring the `! .Method.GetServerStreaming()` version of this function as something like this?
```
func request_{{.Method.Service.GetName}}_{{.Method.GetName}}_{{.Index}}(ctx context.Context, client {{.Method.Service.GetName}}Client, req *http.Request, pathParams map[string]string) (proto.Message, runtime.ServerMetadata, error)
```
With this approach `forward_xxx` function does not need to care about `.GetClientStreaming`.